src/routing/route.cc

```diff
@@ -190,31 +200,164 @@ LaneSRangeRelation Route::ComputeLaneSRangeRelation(const api::LaneSRange& lane_
-    return s1_is_in_range && s0_is_in_range;
+api::LaneSRoute Route::ComputeLaneSRoute(const api::RoadPosition& start_position) const {
+  MALIPUT_THROW_UNLESS(start_position.lane != nullptr);
+
+  const api::RoadPosition& end_road_position = end_route_position();
+  // Creates a point like api::LaneSRange to find the appropriate LaneSRangeIndex of the start
+  // and end positions.
+  const api::LaneSRange end_lane_s_range_point(end_road_position.lane->id(),
+                                               api::SRange(end_road_position.pos.s(), end_road_position.pos.s()));
+  const api::LaneSRange start_lane_s_range_point(start_position.lane->id(),
+                                                 api::SRange(start_position.pos.s(), start_position.pos.s()));
+
+  const std::optional<LaneSRangeIndex> start_lane_s_range_index = FindLaneSRangeIndex(start_lane_s_range_point);
+  MALIPUT_THROW_UNLESS(start_lane_s_range_index.has_value());
+
+  const std::optional<LaneSRangeIndex> end_lane_s_range_index = FindLaneSRangeIndex(end_lane_s_range_point);
+  MALIPUT_THROW_UNLESS(end_lane_s_range_index.has_value());
+
+  // Inserts @p lane_s_range_index into @p lane_s_range at the beginning.
+  auto prepend_lane_s_range_to_route = [this](const LaneSRangeIndex& lane_s_range_index,
+                                              std::vector<api::LaneSRange>* lane_s_ranges) -> void {
+    lane_s_ranges->insert(lane_s_ranges->begin(), GetLaneSRange(lane_s_range_index.first, lane_s_range_index.second));
   };
 
-  for (size_t i = 0; i < phases_.size(); ++i) {
-    for (size_t j = 0; j < phases_[i].lane_s_ranges().size(); ++j) {
-      if (is_lane_s_range_contained(phases_[i].lane_s_ranges()[j], lane_s_range)) {
-        return {std::make_pair(i, j)};
+  // Returns true when @p lane_s_range_index_a.first is equal to @p lane_s_range_index_b.first.
+  auto is_same_phase = [](const Route::LaneSRangeIndex& lane_s_range_index_a,
+                          const Route::LaneSRangeIndex& lane_s_range_index_b) -> bool {
+    return lane_s_range_index_a.first == lane_s_range_index_b.first;
+  };
+
+  LaneSRangeIndex current_lane_s_range_index = end_lane_s_range_index.value();
+  std::vector<api::LaneSRange> lane_s_ranges;
+
+  // The following double nested loop performs the routing within the Route from the ending api::LaneSRange
+  // towards the starting api::LaneSRange. The is always a possible logical solution, and it is based on
+  // one simple set of premises:
+  //
+  // 1- Move straight backwards first, otherwise
+  // 2- Move towards the left / right, the one that offers the lesser number of hops with a possible
+  //    straight backwards movement. Go to step 1.
+
+  // TODO(#543): this process should be simpler once we process in the constructor end-to-end Phase
+  // connectivity as we can cache in the interfaces those LaneSRangeIndices that connect with each other.
+  while (!is_same_phase(current_lane_s_range_index, *start_lane_s_range_index)) {
```

Comment:
Is there a chance that this will loop forever? The body of this while loop only considers the straight predecessor, not the left / right. How does it satisfy step (2) above (lines 243-244)?

Reply:
> Is there a chance that this will loop forever? 

I don't think so, but I've constrained the loops with index counters to make sure it does not loop forever.

> The body of this while loop only considers the straight predecessor, not the left / right.

Correct, because there is no path contiguity with preceding left/right and succeeding left / right api::LaneSRanges. Please take a look at the following picture:

![image](https://github.com/maliput/maliput/assets/3825465/d7e3ee12-7410-4716-98e0-c9e02143f389)

The table below depicts the relationships between the center Phase's api::LaneSRange (Phase:0, api::LaneSRange: 1) and the previous and posterior api::LaneSRanges. As you can see there is no traversable path through only the left and right preceding and suceeding api::LaneSRanges. The only possible traversable path requires going through a straight preceding / suceeding api::LaneSRange.

The circles in the path represent a hop in the index of the api::LaneSRange under evaluation from the end till the beginning.

About the step 2 in my comment:

> 2- Move towards the left / right, the one that offers the lesser number of hops with a possible
> straight backwards movement. Go to step 1.

In this example I am not offering a sample route that evaluates that, but it is the responsibility of `FindStraightPredecessor()` to find the left or right sense of lateral displacement.